p2panda-rs/src/storage_provider/traits/storage_provider.rs

```diff
@@ -0,0 +1,484 @@
+        let entry_with_operation = EntryWithOperation::new(
+            params.entry_encoded().to_owned(),
+            params.operation_encoded().to_owned(),
+        )?;
+
+        // Decode author, entry and operation. This conversion validates the operation hash
+        let author = params.entry_encoded().author();
+        let entry_encoded = params.entry_encoded();
+        let entry = decode_entry(params.entry_encoded(), Some(params.operation_encoded()))?;
+        let operation = Operation::from(params.operation_encoded());
+
+        // Every operation refers to a document we need to determine. A document is identified by the
+        // hash of its first `CREATE` operation, it is the root operation of every document graph
+        let document_id = if operation.is_create() {
+            // This is easy: We just use the entry hash directly to determine the document id
+            DocumentId::new(entry_encoded.hash().into())
+        } else {
+            // For any other operations which followed after creation we need to either walk the operation
+            // graph back to its `CREATE` operation or more easily look up the database since we keep track
+            // of all log ids and documents there.
+            //
+            // We can determine the used document hash by looking at what we know about the previous
+            // entry in this author's log.
+            //
+            // @TODO: This currently looks at the backlink, in the future we want to use
+            // "previousOperation", since in a multi-writer setting there might be no backlink for
+            // update operations! See: https://github.com/p2panda/aquadoggo/issues/49
+            let backlink_entry_hash = entry
+                .backlink_hash()
+                .ok_or(PublishEntryError::OperationWithoutBacklink)?;
+
+            self.get_document_by_entry(backlink_entry_hash)
+                .await?
+                .ok_or(PublishEntryError::DocumentMissing)?
+        };
+
+        // Determine expected log id for new entry
+        let document_log_id = self
+            .find_document_log_id(&author, Some(&document_id))
+            .await?;
+
+        // Check if provided log id matches expected log id
+        if &document_log_id != entry.log_id() {
+            return Err(PublishEntryError::InvalidLogId(
+                entry.log_id().as_u64(),
+                document_log_id.as_u64(),
+            )
+            .into());
+        }
+
+        // Get related bamboo backlink and skiplink entries
+        let entry_backlink_bytes = if !entry.seq_num().is_first() {
+            self.entry_at_seq_num(&author, entry.log_id(), &entry.seq_num_backlink().unwrap())
+                .await?
+                .map(|link| {
+                    let bytes = link.entry_encoded().to_bytes();
+                    Some(bytes)
+                })
+                .ok_or(PublishEntryError::BacklinkMissing)
```

Comment:
I don't think this error would ever be triggered as the one above in line 137 would always be tripped first.